post/2023/beyond-sf/index.qmd

```diff
@@ -0,0 +1,305 @@
+3. We pass the list of `sfg` objects to create an `sfc`
+
+We can included this in a data frame 
+
+```{r}
+library(dplyr, warn.conflicts = FALSE)
+
+dmnd <- diamonds |> 
+  select(price, clarity) |> 
+  bind_cols(geometry = pnts)
+
+head(dmnd)
+```
+
+Now we have a tibble with price, clarity, and geometry! Huge win! What if we wanted to calculate the average price by clarity and keep the geometries?
+
+```{r}
+dmnd |> 
+  group_by(clarity) |> 
+  summarise(avg_price = mean(price))
+```
+
+Unfortunately, we lose the geometry just like we would for any other column that wasn't included in the `summarise()` call. To keep it, we would need to perform an operation on the geometry itself. 
+
+```{r}
+dmnd |> 
+  group_by(clarity) |> 
+  summarise(
+    avg_price = mean(price),
+    geometry = st_union(geometry)
+  )
+```
+
+Wouldn't it be nice if the geometry knew to do that? Well, that is exactly what `sf` objects are.
+
+## `sf` objects
+
+`sf` objects are just data frames with a geometry column that is *sticky* and *smart.* We can create an `sf` object if an `sfc` column is present in a data frame by using `st_as_sf()`. 
+
+```{r}
+dmnd_sf <- st_as_sf(dmnd)
+```
+
+Doing this creates an object of class `sf`. The two things that make an `sf` object so special are the class `sf` and the attribute `sf_column`. 
+
+```{r}
+attr(dmnd_sf, "sf_column") # <1>
+```
+1. `attr()` lets us access a single attribute by name
+This attribute tells us which column is the geometry. Because we have this `sf` can implement its own methods for common functions like `select()`, `mutate()`, `aggregate()`, `group_by()`, etc which always keep the `attr(x, "sf_column")` attached to the data frame. 
+Having the class and attribute allow methods like `summarise()` to be written for the class itself and handle the things like unioning geometry for us. 
+
+```{r}
+dmnd_sf |> 
+  group_by(clarity) |> 
+  summarise(avg_price = mean(price))
+```
+Note that this is just like what we wrote earlier except that we didn't have to handle the geometry column manually. If we compare these two approaches and ignore attribute difference like `sf_column` and `class` we can see that they are identical.
+
+```{r}
+x <- dmnd |> 
+  group_by(clarity) |> 
+  summarise(
+    avg_price = mean(price),
+    geometry = st_union(geometry)
+  )
+
+y <- dmnd_sf |> 
+  group_by(clarity) |> 
+  summarise(avg_price = mean(price))
+
+all.equal(x, y, check.attributes = FALSE) # <1>
+```
+1. We ignore attributes because they _will_ be different because the classes are different
+
+## Other geometry vectors
+
+Knowing how `sf` works can be helpful for understanding how we can ease our reliance on it for all geospatial operations. Instead of thinking of the entire sf data frame as the thing that handles all geometry operations we now know that it is the `sfc` geometry column. 
```

Comment:
```suggestion
Knowing how **sf** works can be helpful for understanding how we can ease our reliance on it for all geospatial operations. Instead of thinking of the entire **sf** data frame as the thing that handles all geometry operations we now know that it is the `sfc` geometry column. 
```